libraries/AP_LTM_Telem/AP_LTM_Telem.cpp

```diff
@@ -0,0 +1,233 @@
+    lt_buff[1] = 0x54; //T
+    // FRAMEID
+    lt_buff[2] = 0x47; // G ( gps frame at 5hz )
+    // PAYLOAD
+    lt_buff[3] = (_uav.lat >> 8 * 0) & 0xFF;
+    lt_buff[4] = (_uav.lat >> 8 * 1) & 0xFF;
+    lt_buff[5] = (_uav.lat >> 8 * 2) & 0xFF;
+    lt_buff[6] = (_uav.lat >> 8 * 3) & 0xFF;
+    lt_buff[7] = (_uav.lon >> 8 * 0) & 0xFF;
+    lt_buff[8] = (_uav.lon >> 8 * 1) & 0xFF;
+    lt_buff[9] = (_uav.lon >> 8 * 2) & 0xFF;
+    lt_buff[10] = (_uav.lon >> 8 * 3) & 0xFF;
+    lt_buff[11] = (_uav.gndspeed >> 8 * 0) & 0xFF;
+    lt_buff[12] = (_uav.alt >> 8 * 0) & 0xFF;
+    lt_buff[13] = (_uav.alt >> 8 * 1) & 0xFF;
+    lt_buff[14] = (_uav.alt >> 8 * 2) & 0xFF;
+    lt_buff[15] = (_uav.alt >> 8 * 3) & 0xFF;
+    lt_buff[16] = ((_uav.sats_visible << 2) & 0xFF)
+            | ((_uav.fix_type > 3 ? 3 : _uav.fix_type) & 0b00000011); // last 6 bits: sats number, first 2:fix type (0, 1, 2, 3)
+    send_LTM(lt_buff, LTM_GFRAME_SIZE);
+    _ltm_scheduler++;
+}
+
+// Sensors frame
+void AP_LTM_Telem::generate_Sframe(void)
+{
+    get_sense();
+
+    uint8_t lt_buff[LTM_SFRAME_SIZE];
+    // START
+    lt_buff[0] = 0x24; //$
+    lt_buff[1] = 0x54; //T
+    // FRAMEID
+    lt_buff[2] = 0x53; //S 
+    // PAYLOAD
+    lt_buff[3] = (_uav.volt >> 8 * 0) & 0xFF; // VBAT converted to mV
+    lt_buff[4] = (_uav.volt >> 8 * 1) & 0xFF;
+    lt_buff[5] = (_uav.amp >> 8 * 0) & 0xFF; // actual current instead of consumed mAh in regular LTM
+    lt_buff[6] = (_uav.amp >> 8 * 1) & 0xFF;
+    lt_buff[7] = (_uav.rssi >> 8 * 0) & 0xFF;
+    lt_buff[8] = (_uav.airspeed >> 8 * 0) & 0xFF;
+    lt_buff[9] = ((_uav.flightmode << 2) & 0xFF)
+            | ((_uav.failsafe << 1) & 0b00000010) | (_uav.armstat & 0b00000001); // last 6 bits: flight mode, 2nd bit: failsafe, 1st bit: arm status.
+    send_LTM(lt_buff, LTM_SFRAME_SIZE);
+    _ltm_scheduler++;
+}
+
+// Attitude frame
+void AP_LTM_Telem::generate_Aframe(void)
+{
+    get_attitude();
+
+    uint8_t lt_buff[LTM_AFRAME_SIZE];
+    // A Frame: $T(2 bytes)A(1byte)PITCH(2 bytes)ROLL(2bytes)HEADING(2bytes)CRC(xor,1byte)
+    // START
+    lt_buff[0] = 0x24; //$
+    lt_buff[1] = 0x54; //T
+    // FRAMEID
+    lt_buff[2] = 0x41; //A 
+    // PAYLOAD
+    lt_buff[3] = (_uav.pitch >> 8 * 0) & 0xFF;
+    lt_buff[4] = (_uav.pitch >> 8 * 1) & 0xFF;
+    lt_buff[5] = (_uav.roll >> 8 * 0) & 0xFF;
+    lt_buff[6] = (_uav.roll >> 8 * 1) & 0xFF;
+    lt_buff[7] = (_uav.heading >> 8 * 0) & 0xFF;
+    lt_buff[8] = (_uav.heading >> 8 * 1) & 0xFF;
+    send_LTM(lt_buff, LTM_AFRAME_SIZE);
+    _ltm_scheduler++;
+}
+
+// Send LTM
+void AP_LTM_Telem::generate_LTM(void)
+{
+    if (_ltm_scheduler & 1) {   // is odd
+        generate_Aframe();
+    } else {                    // is even
+        if (_ltm_scheduler % 4 == 0) {
+            generate_Sframe();
+        } else {
+            generate_Gframe();
+        }
+    }
+    if (_ltm_scheduler > 9) {
+        _ltm_scheduler = 1;
+    }
+}
+
+void AP_LTM_Telem::get_position(void)

```

Comment:
Any particular reason to store all of this state?  Since you're refreshing for every packet you may as well not pass the data back via instance members...

Same comment applies to attitude and sensor frame data.

Reply:
Rewritten